### PR TITLE
Vectorize no_local GEMM kernel

### PR DESCRIPTION
**Pull request by CodeplayBen**
This PR vectorizes the no_local memory version of the GEMM kernel. To avoid issues with both A and B not being vectorizable in all cases due to transposition, in contrast the kernel now loads multiple k of values when vectorized - A loads a full block of values while B only loads  a partial block to reduce register pressure. For each partial block the partial result is calculated then after all the partials are calculated the results block is stored as normal.

### CODE REVIEW

File: src/operations/blas3/gemm_no_local.hpp [hunk 1]
```diff
@@ -394,76 +415,444 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
+   * checking.
+   * @param in_row : function which checks the boundary in the row direction.
+   * @param in_col : function which checks the boundary in the col direction.
+   * @param out_of_range: exits the function early if block is out of range.
+   */
+  template <bool check_row, bool check_col, index_t work_per_load, bool trans,
+            typename PointerType, typename RowCheckType, typename ColCheckType>
+  SYCL_BLAS_INLINE typename std::enable_if<!trans>::type load_single_b(
+      PointerType ptr, element_t *reg, const index_t &row_ofs,
+      const index_t &col_ofs, const RowCheckType &in_row,
```
Comment (jwlawson) [inline src/operations/blas3/gemm_no_local.hpp]: Could `in_row` and `in_col` have better names? `is_valid_row` or something?

File: src/operations/blas3/gemm_no_local.hpp [hunk 2]
```diff
@@ -394,76 +415,444 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
+          // if in range perform a vectorised load
+          in_vec.template load<address_t::global_space>(0, ptr + j * ptr_next);
+        } else {
+          // if not in range perform element-wise load checking boundaries at
+          // each load.
+#pragma unroll
+          for (int l = 0; l < work_per_load; l++) {
+            if (do_check<check_row>(in_row(l)) &&
+                do_check<check_col>(in_col(i))) {
+              reinterpret_cast<element_t *>(&in_vec)[l] =
```
Comment (jwlawson) [inline src/operations/blas3/gemm_no_local.hpp]: I'm not 100% sure this is safe. Unfortunately SYCL doesn't provide a programmatic way to set vector elements. I don't think this should block this PR, but we could think about adding something like the helpers used in SYCL-DNN for this: https://github.com/codeplaysoftware/SYCL-DNN/blob/master/src/helpers/vector_element.h

File: src/operations/blas3/gemm_load_store.hpp [hunk 3]
```diff
@@ -28,38 +28,48 @@
...
-  /*! @brief Performs a coalesced non-vectorised load when the current block is
+  /*! @brief Performs a coalesced non-vectorized load when the current block is
    * not internal.
    * @tparam trans Whether the source matrix is transposed or not.
    * @tparam internal True if the current block is internal and no bounds
    * checking is required.
    * @tparam ld The leading dimension of the destination memory.
    */
+
   template <bool trans, bool internal, int ld, typename packet_t = PacketType,
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_load_store.hpp]: Yes I think it was just a leftover parameter from a previous iteration, fixed!

File: src/operations/blas3/gemm_no_local.hpp [hunk 4]
```diff
@@ -153,6 +167,8 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
   SYCL_BLAS_INLINE void eval(cl::sycl::nd_item<1> id) noexcept {
     index_t m = a_.get_size_row();
     index_t n = b_.get_size_col();
+    const index_t mc = m;
+    const index_t nc = n;
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Have renamed them to original_m/n to reflect that they are the unaltered m and n values (m and n are subsequently altered).

File: src/operations/blas3/gemm_no_local.hpp [hunk 5]
```diff
@@ -186,20 +202,39 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
-    /* work item id per row */
-    const index_t local_item_id_row = item_id % wg_rows;
-    /* work item id per column */
-    const index_t local_item_id_col = item_id / wg_rows;
     /* the start position of the tile-row per work group */
     const index_t wg_row = tile_id_row * block_rows;
     /* the start position of the tile-column per work group */
     const index_t wg_col = tile_id_col * block_cols;
+    /*!
+     * @brief is_internal_block_m and is_internal_block_n is used to distinguish
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Updated! This was leftover from when they were separate bools.

File: src/operations/blas3/gemm_no_local.hpp (hunk 1 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Changed! Thanks for the suggestion.

File: src/operations/blas3/gemm_no_local.hpp (hunk 2 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: I agree it probably needs further discussion. Helpers could work assuming there is no performance penalty.

File: src/operations/blas3/gemm_no_local.hpp [hunk 6]
```diff
@@ -394,76 +415,453 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
+   * @param boundary_check_n : Function which checks boundary in N dimension.
+   * @param A_ptr_index : index of next element for A
+   * @param B_ptr_index : index of next element for B
+   * @param lda : leading dimension of A
+   * @param ldb : leading dimension of B
+   * @param k : the current value of K from the main loop which calls this
+   * method.
+   * @param reg_a : Pointer to private register for A.
+   * @param reg_b : Pointer to private register for B.
+   * @param reg_res : Pointer to private register for partial result.
```
Comment (mehdi-goli) [inline src/operations/blas3/gemm_no_local.hpp]: Same here, reg_res should be the result not partial

File: src/operations/blas3/gemm_no_local.hpp [hunk 7]
```diff
@@ -394,76 +415,453 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
+            B + ofs, reg_b, j, col_ofs,
+            [=](const index_t &idx) SYCL_BLAS_ALWAYS_INLINE { return idx < k; },
+            boundary_check_n, out_of_range);
+
+        /*
+         * Computing a partial GEMM for the loaded block of reg_a and partial
+         * reg_b and adding the result into reg_res
+         */
+
+        compute_block_gemm_no_shared<packet_size>(i + j, reg_a, reg_b, reg_res);
```
Comment (mehdi-goli) [inline src/operations/blas3/gemm_no_local.hpp]: the iteration type is const Index& so (i + j) force compiler to allocate memory and pass the address of that memory to the function, There is no need to make an Index to be a reference

Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local.hpp (hunk 6 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local.hpp [hunk 8]
```diff
@@ -304,69 +333,61 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
+    value_t reg_a[item_rows * packet_size];
+    /* temporary register used to prefetch elements of B*/
+    value_t reg_b[packet_size];
     do {
       auto A = orig_A;
       auto B = orig_B;
       auto C = orig_C;
 
-      /* 2D register array used to store the result C*/
+      /* register array used to store the partial result*/
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: No you are correct, it is not a partial result. Fixed!

File: src/operations/blas3/gemm_no_local.hpp [hunk 9]
```diff
@@ -210,57 +245,42 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
-    };
-    const auto boundary_check_n = [&](index_t dim_n_b_start) {
-      return dim_n_b_start < n;
-    };
-    const auto boundary_check_c = [&](index_t dim_m_c_start,
-                                      index_t dim_n_c_start) {
-      return (dim_m_c_start < m && dim_n_c_start < n);
+    const auto boundary_check_m = [&](const index_t &idx) { return idx < m; };
+    const auto boundary_check_n = [&](const index_t &idx) { return idx < n; };
+    const auto boundary_check_c = [&](const index_t &dim_m_c_start,
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local.hpp [hunk 10]
```diff
@@ -210,57 +245,42 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
-      return dim_m_a_start < m;
-    };
-    const auto boundary_check_n = [&](index_t dim_n_b_start) {
-      return dim_n_b_start < n;
-    };
-    const auto boundary_check_c = [&](index_t dim_m_c_start,
-                                      index_t dim_n_c_start) {
-      return (dim_m_c_start < m && dim_n_c_start < n);
+    const auto boundary_check_m = [&](const index_t &idx) { return idx < m; };
+    const auto boundary_check_n = [&](const index_t &idx) { return idx < n; };
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local.hpp [hunk 11]
```diff
@@ -210,57 +245,42 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
-    const auto boundary_check_m = [&](index_t dim_m_a_start) {
-      return dim_m_a_start < m;
-    };
-    const auto boundary_check_n = [&](index_t dim_n_b_start) {
-      return dim_n_b_start < n;
-    };
-    const auto boundary_check_c = [&](index_t dim_m_c_start,
-                                      index_t dim_n_c_start) {
-      return (dim_m_c_start < m && dim_n_c_start < n);
+    const auto boundary_check_m = [&](const index_t &idx) { return idx < m; };
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local.hpp [hunk 12]
```diff
@@ -186,20 +202,39 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, tile_type,
...
+     * @brief is_internal_block is used to distinguish
+     * the internal block. Therefore, work items using these blocks don't need to
+     * check for boundaries. Checking the packet size is a workaround because
+     * normally the vector size and item rows/cols must all be equal, but when
+     * vectorization is disabled the vector size is always 1 and the algorithm
+     * breaks.
+     */
+    const bool is_internal_block =
+        packetize_t::packet_size !=
+            1 &&  // This is a workaround when vectorization is disabled.
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local.hpp]: Fixed!

File: src/operations/blas3/gemm_load_store.hpp [hunk 13]
```diff
@@ -28,38 +28,47 @@
...
+  template <index_t dimension>
+  static constexpr bool check_size() {
+    return dimension == packet_size;
+  }
 #else
   // In the case where vectorization is not enabled, always set to 1
   using PacketType = cl::sycl::vec<value_t, 1>;
   static constexpr size_t packet_size = 1;
+  template <index_t dimension>
+  static constexpr bool check_size() {
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_load_store.hpp]: Fixed!

File: src/operations/blas3/gemm_load_store.hpp [hunk 14]
```diff
@@ -28,38 +28,47 @@
...
+supported).
 */
-template <size_t vector_size, typename value_t>
+template <size_t vector_size, typename value_t, typename index_t>
 struct Packetize {
 #ifdef GEMM_VECTORIZATION_SUPPORT
   using PacketType = cl::sycl::vec<value_t, vector_size>;
   static constexpr size_t packet_size = vector_size;
+  template <index_t dimension>
+  static constexpr bool check_size() {
```
[the remark this answers is not in the document]
Comment (CodeplayBen) [reply src/operations/blas3/gemm_load_store.hpp]: Fixed!

**Comment by CodeplayBen** [comment]
Going to close this and open a new pull request based on the changes in #230. Feedback from comments here has been addressed and will be carried to the new pull request.

**Pull request closed without merging**